### ISSUE TITLE
Request for suggestions: Make it easier to `append` and `prepend` items in arrays

### ISSUE THREAD

**Issue by karreiro** [Shopify/liquid#1851]
**Author**: Shopify
**Expected end date**: December 5, 2024

### Background

Handling arrays in Liquid is challenging.

Developers often need to convert arrays into strings, perform operations, and then split them back, relying heavily on the Liquid [String API](https://shopify.dev/docs/api/liquid/filters/string-filters) as it's more robust than the [Array API](https://shopify.dev/docs/api/liquid/filters/array-filters).

```liquid
{% assign titles = collections.last.products | map: "title" %}

{{
  titles
  | join: ","
  | append: ",default"
  | split: ","
}}
{{
  titles
  | join: ","
  | prepend: "All,"
  | split: ","
}}
```

### Proposal

We should eliminate workarounds like string concatenation and splitting and make it easier to perform common tasks such as adding new items in an array. This will make Liquid templates simpler, less verbose, and more declarative.

To make that happen, we should support the `append` and `prepend` filters in arrays. They will return a new array with the specified element added to the end or beginning, respectively.

```liquid
{% assign titles = collections.last.products | map: "title" %}

{{ titles | append: "default" }}
{{ titles | prepend: "All" }}
```

### Limitations

This proposal is not backward compatible [[1]](https://github.com/Shopify/liquid/blob/42b6763546f218dc8ad1184e10c5e2d9a262b006/lib/liquid/standardfilters.rb#L621-L623) [[2]](https://github.com/Shopify/liquid/blob/42b6763546f218dc8ad1184e10c5e2d9a262b006/lib/liquid/standardfilters.rb#L650-L652).

```liquid
{{ titles | append: "new" }}
{% # => "[\"Example T-Shirt\", \"Classic Varsity Top\", \"Classic Leather Jacket\", \"Dark Denim Top\"]new" %}
```

Therefore, the implementation is pending an impact assessment. If we identify it's impossible to handle backward compatibility, we aim to introduce only the `append` functionality instead, using an unused filter name (e.g. `add`). 

We're proposing the `append`/`prepend` direction first because we consider this a more idiomatic and cohesive direction for the Liquid API.

### Call for suggestions

We welcome any feedback or opinions on this proposal. Please share your thoughts by December 5, 2024. Your input is valuable as we prepare to begin active development on this initiative.

**Comment by galenking** [comment]
Would lack of backwards-compatibility not break any existing theme that’s treating append as a string? Will there be any way to determine which version of liquid a theme uses?

Would it not make more sense to use `push` and `unshift`? Although, I personally find `prepend` and `append` more intuitive and I do think it’s a better pairing for the next generation of devs.

**Comment by jg-rp** [comment]
> Would it not make more sense to use `push` and `unshift`?

I think I prefer `push` and `push_left`.

**Comment by kinngh** [comment]
`append` and `prepend` are easy to confuse with string concatenation. Keeping it to a simple `push` would do a great job.

Adding element to the end of array:

```liquid
{% assign titles = titles | append: 'new_title'  %}

OR

{{ titles | append: 'new_title' }}
```

Adding element at the beginning of an array:

```liquid
{% assign titles = 'new_title' | append: titles %}
```

**Comment by wbelk** [comment]
+1 on arrays

better support for an array type

```
{% assign foo = [ 'c' ] %}
# foo: ['c']

{% push foo = 'a' %}
# foo: ['c', 'a']

{% unshift foo = 'b' %}
# foo: ['b', 'c', 'a']

{{ foo | pop }}
# 'a'

{{ foo | shift }}
# 'b'
```

**Comment by davidhollenbeckx** [comment]
would love better array support. using existing keywords doesnt make sense to me. the functionality is different, a different keyword makes sense. plus as mentioned backwards compatibility issues would cause massive problems for custom dev shops

**Comment by jg-rp** [comment]
> we did consider the pop, push, shift, and unshift names

I agree, `prepend` and `append` would be better than `push` etc., if you can accept the behavioural change. Although `append` might still imply a mutating operation to some (Ruby's `Array.append` is an alias for `Array.push`, Python's `List.append` mutates the list).

Modifying the `plus` filter to handle arrays might better convey the fact that you're getting a new array, but I'm not convinced it is the best choice.
